### DISCUSSION TITLE
SensitivityFilter in SIMP Optimization

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#25989] [Q&A Modules: General] [answered]
Hello,

I started exploring the SIMP omptimization capabilities of MOOSE recently and noticed that the `SensitivityFilter` only partially behaves the way I'd expect it to or rather how I'm used to. It does prevent checkerboarding and introduces a minimum length scale dependent on the filter radius - as expected. However the boundaries are way clearer than I'd anticipate. For reference, here are two pictures from an optimization with similar setup from ABAQUS and MOOSE with a filter radius of 10 times the element size:

![Bildschirmfoto von 2023-11-07 12-26-44](https://github.com/idaholab/moose/assets/98457971/caaa50e6-161c-4939-b71c-3250fd1679c3)
![Bildschirmfoto von 2023-11-07 12-26-30](https://github.com/idaholab/moose/assets/98457971/5fae3b1e-544e-422e-aeae-178a21fc8dc8)

The clearer boundaries from the MOOSE result are instantly noticeable. A straight 0-1-design is considered a good thing, I was just wondering how exactly the MOOSE `SensitivityFilter` works since I recon it differs from the standard sensitivity filter used [here](https://link.springer.com/article/10.1007/s00158-010-0594-7) for example. I'm primarily having a hard time understanding the functionality of `RadialAverage` and the meaning of `_JxW[gp]`. Maybe the resulting MOOSE filter is one of the ones mentioned [here](https://link.springer.com/article/10.1007/s00158-006-0087-x) in section 5.2? 

Thanks in advance.

**Comment by maxnezdyur** [comment, marked as answer]
There are multiple sensitivity filter forms with different pros and cons. When using the Radial Average Sensitivity filter, the current psuedo density is not used in the calculation, but is used in the paper you provided. There is no "correct" filter since this is mostly a heuristic used to alleviate checkerboarding and have minimum length scale.

**Reply by Flolaffel** [reply]
Thanks for your answer. I failed to think about the absence of the density in the MOOSE filter.

I'm aware there is not one correct sensitivity filter as the second paper I linked suggests. If one aims to replicate a similar filter to the one used in the widely known [88 line Matlab code](https://link.springer.com/article/10.1007/s00158-010-0594-7) they can use a `ParsedMaterial` for the numerator which is passed to `RadialAverage` in conjunction with a slight modification to the denominator in `SensitivityFilter` (dividing by the pseudo density of the current element). See `2d_mbb_sens_filter.i` in my simp repo for reference.

```
[filter_mat]
    type = ParsedMaterial
    property_name = filter_mat
    material_property_names = 'sensitivity'
    coupled_variables = 'mat_den'
    expression = 'sensitivity*mat_den'
  []
```

**Reply by maxnezdyur** [reply]
Very cool, do you have a picture by chance using your method?

**Reply by Flolaffel** [reply]
Sure!

**Standard MOOSE filter method**
![moose_mbb](https://github.com/idaholab/moose/assets/98457971/aa3e5315-cdd9-4936-ba8e-fc5aa1f824ed)


**Modified MOOSE filter**
![moose_mbb_top88](https://github.com/idaholab/moose/assets/98457971/dee97155-8832-445f-8a88-8ef5cd895dc4)


**Matlab 88 line code result**
![matlab_mbb_top88](https://github.com/idaholab/moose/assets/98457971/db3202f8-ce27-4cfb-be2a-a7f95cf2caab)


Please note that the modified MOOSE filter is not exactly the same since I haven't figured out how to pass the `weight` (or rather a `weight_sum`) variable from `ThreadedRadialAverageLoop.C` to `SensitivityFilter.C` yet. So that is still missing in the denominator of the filter.
